### PR TITLE
feat(android): Add support for AGP 8 in example, add compileOptions to build.gradle

### PR DESCRIPTION
**Pull request by Skyost**
# Description

This PR adds support of namespace property to support Android Gradle Plugin (AGP) 8. Projects with AGP < 4.2 will not supported anymore. It fixes the following error :

Namespace error

See flutter/flutter#125181.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:`, `chore:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation and added dartdoc comments with `///`, where necessary.
- [x] I have updated/added relevant examples in [example].

## Breaking Change

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

### Migration instructions

You can (and you should !) upgrade your Android project to use AGP 8. In your `build.gradle` :

```gradle
buildscript {
  // ...

  dependencies {
    classpath 'com.android.tools.build:gradle:8.0.1'
    // ...
  }
}
```

Also don't forget to add a namespace in the `android` section of `app/build.gradle` :

```gradle
android {
  // ...
  namespace 'your.namespace'
}
```

## Related Issues

* https://github.com/flutter/flutter/issues/125621#issuecomment-1525995461
* https://github.com/flutter/flutter/issues/125181

[issue database]: https://github.com/bluefireteam/audioplayers/issues
[Contributor Guide]: https://github.com/bluefireteam/audioplayers/blob/main/contributing.md#feature-requests--prs
[Conventional Commit]: https://conventionalcommits.org
[example]: https://github.com/bluefireteam/audioplayers/tree/main/packages/audioplayers/example

### CODE REVIEW

**Comment by Skyost** [comment]
Note : there's a way to still support AGP 4.2. All we have to do is :

```gradle
if (project.android.hasProperty("namespace")) {
  namespace 'xyz.luan.audioplayers'
}
```

But I'm not sure that it's the _best_ idea as AGP 4.2 is old. As you want, I can adapt my PR if you ask to 🙂

**Comment by Gustl22** [comment]
As far as I remember the namespace feature didn't work with `flutter < [3.7.12](https://github.com/flutter/flutter/releases/tag/3.7.12)` (after clearing the gradle cache). Do we have to raise the minimum Flutter version in pubspec.yaml in order to make it working?

**Comment by Skyost** [comment]
Not needed, because we're checking if `namespace` is available first :

```gradle
if (project.android.hasProperty('namespace')) {
  namespace 'xyz.luan.audioplayers'
}
```

**Comment by Skyost** [comment]
@vbuberen That's what I did first, but I reverted it in https://github.com/bluefireteam/audioplayers/pull/1503/commits/42d3150ee040e4cf5399c74e2e31ac096121ce9d as I thought it was not my role to take such a decision.

**Comment by cian-bayer** [comment]
Hi, any eta on when this will be merged and released? I just upgraded my flutter project to AGP 8 and I believe audioplayers missing the `namespace` parameter is the last thing blocking my project from building for android. Thanks

**Comment by Skyost** [comment]
@cian-bayer Feel free to use the fork. In your `pubspec.yaml` :

```yaml
audioplayers:
  git:
    url: https://github.com/Skyost/audioplayers.git
    path: packages/audioplayers/
```

File: packages/audioplayers/example/android/gradle.properties [hunk 1]
```diff
@@ -1,3 +1,6 @@
 org.gradle.jvmargs=-Xmx1536M
 android.useAndroidX=true
 android.enableJetifier=true
+android.defaults.buildfeatures.buildconfig=true
+android.nonTransitiveRClass=false
+android.nonFinalResIds=false
```
Comment (Gustl22) [inline packages/audioplayers/example/android/gradle.properties:6]: Are these changes needed? According to the [release notes](https://developer.android.com/build/releases/gradle-plugin#default-changes) the defaults are changed to improve build performance.

Comment (Skyost) [reply packages/audioplayers/example/android/gradle.properties:6]: They're not. In fact, they happened when I used the Android Studio migration tool. I can remove them if you want.

**Comment by vbuberen** [comment]
> Changing AGP version fails, due to [flutter/flutter@3a6c4eb](https://github.com/flutter/flutter/commit/3a6c4eb14d778f6efba4c5813425fb6145b296b5) is not released yet. Then we also need to move to `flutter 3.10.0` first, which depends on [flame-engine/flame@b41622d](https://github.com/flame-engine/flame/commit/b41622db8faa7559328f83f8f1d93ec4c6386961) to be released.

There is no need to bump to AGP 8.0 with this PR. The suggested change is only to add a `namespace` property which is required by projects that already migrated to AGP 8.0, thus, add a compatibility for those that decided to jump on this version already.

File: packages/audioplayers/example/android/gradle.properties (hunk 1 above)
Comment (Gustl22) [reply packages/audioplayers/example/android/gradle.properties:6]: Yes that would be nice 😊

**Comment by Skyost** [comment]
@Gustl22 
> started another PR to consider the AGP only without testing it.

Thanks !

**Comment by vbuberen** [comment]
> Otherwise we are blocked by: [flutter/flutter@3a6c4eb](https://github.com/flutter/flutter/commit/3a6c4eb14d778f6efba4c5813425fb6145b296b5) We then can leave the PR for example open until flutters fix in integration tests is released.
> 
> Edit: already started another PR to consider the AGP only without testing it.

I think there is some misunderstanding from your side on what is needed to support AGP 8. Only adding `namespace` property (which appeared since AGP 4.2 as I shared in one of comments above) is needed and there is no need to bump AGP in the package to 8 for now, like you did in another PR. Thus, there is nothing blocking from adding this support and still having integration tests working.

Here is an example of a similar change, but without bumping to AGP 8: https://github.com/fluttercommunity/plus_plugins/pull/1698/files

**Comment by Gustl22** [comment]
> there is no need to bump AGP in the package to 8 for now, like you did in another PR.

@vbuberen Is there a reason to not upgrade the build gradle tools in `audioplayers_android` package? I think it just works as before.

>Only adding namespace property (which appeared since AGP 4.2 as I shared in one of comments above) is needed

> Here is an example of a similar change, but without bumping to AGP 8

Why should we adding the namespace in the `example` without removing it in the `AndroidManifest.xml`. Then it doesn't have any benefit in my opinion and we should wait changing it, until the integration tests also support the `namespace` property.

**Comment by vbuberen** [comment]
> Why should we adding the namespace in the `example` without removing it in the `AndroidManifest.xml`. Then it doesn't have any benefit in my opinion and we should wait changing it, until the integration tests also support the `namespace` property.

Because adding `namespace` would allow people who already moved their projects to AGP 8 to build their projects fine without errors saying that one of their dependencies is missing `namespace` property. Having package id in AndroidManifest has no effect on `namespace`, so shouldn't be considered as benefit/no benefit, but allows integration tests to work till the `integration_test` package supports the new `namespace` property. 
Thus, having both allows to have a broader compatibility in the end without being blocked because `integration_test` don't understand namespace property as of now.

**Comment by vbuberen** [comment]
> Is there a reason to not upgrade the build gradle tools in `audioplayers_android` package? I think it just works as before.

I had some build issues with AGP 8 back then. Maybe it is not the case anymore as there were quite a few updates to Flutter and tooling in last 3-4 weeks.

**Comment by Gustl22** [comment]
> Because adding namespace would allow people who already moved their projects to AGP 8 to build their projects fine without errors saying that one of their dependencies is missing namespace property. Having package id in AndroidManifest has no effect on namespace, so shouldn't be considered as benefit/no benefit, but allows integration tests to work till the integration_test package supports the new namespace property.

I think we are discussing two different things. I did not mean to change the AndroidManifest in `audioplayers_android`. Changing something in the `example` doesn't have any consequences for the package user / app developer. But `example` is used to do the integration tests. So I propose to nothing change there yet, like done in #1514. 

At some day in the future we should removing the namespace from AndroidManifest in both the example and the audioplayers_android package.

> I had some build issues with AGP 8 back then. Maybe it is not the case anymore as there were quite a few updates to Flutter and tooling in last 3-4 weeks. And my statement in previous comments meant that adding a namespace doesn't require to bump to AGP 8 as well.

I think the AGP version in [audioplayers_android](https://github.com/bluefireteam/audioplayers/blob/415dda3b1621c57ea4b0366187f27f6a189555bf/packages/audioplayers_android/android/build.gradle#LL14C52-L14C52) is overriden by the one of the user's app anyways (here [example](https://github.com/bluefireteam/audioplayers/blob/415dda3b1621c57ea4b0366187f27f6a189555bf/packages/audioplayers/example/android/build.gradle#L9)) so it wouldn't matter bumping it, maybe (?), but leaving it for now as done in [flutter packages](https://github.com/flutter/packages/blob/7d64098a16342b94f4109d14bbd51d395ee74a7d/packages/video_player/video_player_android/android/build.gradle#L12)

**Comment by cian-bayer** [comment]
FWIW, I tried building for android from @Skyost 's branch like this:

```
  audioplayers:
    git:
      url: https://github.com/Skyost/audioplayers.git
      path: packages/audioplayers/
```

and got this error:

```
Launching lib/environment/main_dev/main_dev-local-release.dart on sdk gphone64 arm64 in release mode...
main_dev-local-release.dart:1

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':audioplayers_android'.
> Could not create an instance of type com.android.build.api.variant.impl.LibraryVariantBuilderImpl.
   > Namespace not specified. Please specify a namespace in the module's build.gradle file like so:

     android {
         namespace 'com.example.namespace'
     }

     If the package attribute is specified in the source AndroidManifest.xml, it can be migrated automatically to the namespace value in the build.gradle file using the AGP Upgrade Assistant; please refer to https://developer.android.com/studio/build/agp-upgrade-assistant for more information.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 7s
```

and I made sure to run `flutter pub get`, and my pubspec.lock was pointing to the correct, most recent push in their fork: `ebc4e86293412e93084337b8d8b82ce07687ef1d`

Flutter doctor:

```
XXXX@XXXX project-directory % flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.7.12, on macOS 13.3 22E252 darwin-arm64, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.2)
[✓] VS Code (version 1.78.2)
[✓] Connected device (3 available)
[✓] HTTP Host Availability

• No issues found!
```

**Review by spydon — APPROVED** [review]
